lib/config_cat/config/user_comparator.ex

```diff
@@ -0,0 +1,545 @@
+    with {:ok, user_version} <- to_version(user_value),
+         {:ok, comparison_version} <- to_version(comparison_value) do
+      result = Version.compare(user_version, comparison_version)
+      {:ok, result in valid_comparisons}
+    end
+  end
+
+  defp compare_numbers(user_value, comparison_value, operator) do
+    with {:ok, user_float} <- to_float(user_value),
+         {:ok, comparison_float} <- to_float(comparison_value) do
+      {:ok, operator.(user_float, comparison_float)}
+    end
+  end
+
+  defp compare_datetimes(user_value, comparison_value, valid_comparisons) do
+    with {:ok, user_seconds} <- to_unix_seconds(user_value),
+         {:ok, comparison_seconds} <- to_float(comparison_value) do
+      result =
+        cond do
+          user_seconds < comparison_seconds -> :lt
+          user_seconds > comparison_seconds -> :gt
+          true -> :eq
+        end
+
+      {:ok, result in valid_comparisons}
+    else
+      {:error, :invalid_float} -> {:error, :invalid_datetime}
+      error -> error
+    end
+  end
+
+  defp hash_value(value, %ComparisonContext{} = context) do
+    salted = value <> context.salt <> context.context_salt
+
+    :sha256
+    |> :crypto.hash(salted)
+    |> Base.encode16()
+    |> String.downcase()
+  end
+
+  defp parse_comparison(value) do
+    [length_string, comparison_string] = String.split(value, "_", parts: 2)
+    length = String.trim(length_string)
+    {String.to_integer(length), comparison_string}
+  end
+
+  defp as_text(value, _context) when is_binary(value), do: {:ok, value}
+
+  defp as_text(value, %ComparisonContext{} = context) do
+    %ComparisonContext{condition: condition, key: key} = context
+    attribute_name = UserCondition.comparison_attribute(condition)
+    condition_text = UserCondition.description(condition)
+
+    ConfigCatLogger.warning(
+      "Evaluation of condition (#{condition_text}) for setting '#{key}' may not produce the expected result " <>
+        "(the User.#{attribute_name} attribute is not a string value, thus it was automatically converted to " <>
+        "the string value '#{value}'). Please make sure that using a non-string value was intended.",
+      event_id: 3005
+    )
+
+    user_value_to_string(value)
+  end
+
+  @spec user_value_to_string(Config.value() | DateTime.t() | NaiveDateTime.t() | [String.t()]) ::
+          {:ok, String.t() | nil} | {:error, :invalid_datetime | :invalid_float | :invalid_string_list}
+  def user_value_to_string(nil), do: {:ok, nil}
+
+  def user_value_to_string(%DateTime{} = dt) do
+    with {:ok, seconds} <- to_unix_seconds(dt) do
+      {:ok, to_string(seconds)}
+    end
+  end
+
+  def user_value_to_string(%NaiveDateTime{} = naive) do
+    naive |> DateTime.from_naive!("Etc/UTC") |> user_value_to_string()
+  end
+
+  def user_value_to_string(value) when is_list(value) do
+    with {:ok, list} <- to_string_list(value) do
+      {:ok, inspect(list)}
```

Comment:
Unfortunately, a further issue has been uncovered regarding this conversion: even that matters how the JSON string is formatted.

After discussing this, we had to make an addition to the specification. This explains the why: https://docs.google.com/document/d/1-Vi0YnX57P3gW1-_K_kaR1uZECQowajGJtbGwYvb3a8/edit#heading=h.6ajheae6qud8

Also, I specified a new test to ensure the platform-independent behavior across our SDKs. You can find it here: https://github.com/configcat/common-js/pull/102/commits/f6290509b14ce17740447d5b056c9e972d676a8d

Please make the necessary adjustment and implement this test.

And thank you for your patience, expectedly this will be my last request to you in this PR.

Reply:
Done in b1194b3. This was quite tricky, as Elixir's float formatting is different than JavaScripts and I had to find a way to make them match.